### DISCUSSION TITLE
queryFn returning `undefined` should be cached

### DISCUSSION THREAD

**Discussion by 90dy** [TanStack/query#6029] [General]
Hello there 👋 

When using a query that return `undefined`, the cacheTime and refetch options are not taken into account. It could happen that a call to a server return an undefined value, and we don't want to re-request to prevent too much server-side computations.
We can also want to use useQuery as a better useEffect, that should return nothing.

The workaround is to return something, but it's not really explicit for developer, and is not good for maintainability as not every developer knows about that tricky solution :sob:

If it's to know if a cache exists for that query, react-query should use:

For:
```ts
const queryCache = {
  [queryKey1]: returnValue1,
  [queryKey2]: returnValue2,
}
```
```ts
const isQueryCache1 = queryKey1 in queryCache
```
instead of
```ts
const isQueryCache1 = queryCache[queryKey] === undefined
```

That's a DX issue, so what do you think about that ?

Thanks 🙌

**Comment by TkDodo** [comment]
it's not allowed to return `undefined` from the `queryFn`, because `undefined` is used a kind of special value. for example, when `initialData` is set to `undefined`, it means you don't have `initialData`. If `undefined` were allowed for successful queries, this wouldn't be representable.

There should also be an error in the console in dev mode:

https://github.com/TanStack/query/blob/35539bf34f4e623c9ad70bbcd3d01e1f837bfa63/packages/query-core/src/query.ts#L475-L477

**Reply by 90dy** [reply]
I'm sure there's a workaround, if initialData isn't declared and not assigned, it's not undefined.

```ts
type Options = {
   initialData?: TData
 }
 ```
is not equal to
```ts
type Options = {
  initialData: TData | undefined 
}
```
when `exactOptionalPropertyTypes` is set to `true` in tsconfig.
https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes

If initialData isn't declared and not assigned, `'initialData' in options` will return false.
If initialData is declared but not assigned (eq. `initialData = undefined`), `'initialData' in options` will return true.

**Reply by TkDodo** [reply]
this was just one example of many. we use `undefined` to bail out of state updates with `setQueryData`. We can't do that if undefined can be legally stored as data for successful queries. Basically, we never want to be in this state:

```
status: 'success'
data: undefined
```

because the success status should indicate that we have data, and `undefined` doesn't communicate that at all.

**Reply by 90dy** [reply]
`data` declared means we have data, `undefined` tells us only data is empty.
Sorry I'm not here to bother you but I'm sure we could do something about this and improve the react-query a little more.
Ofc that's not something urgent.
Thank you for your time 🙏

**Reply by TkDodo** [reply]
yes, we're about to improve the typescript types so that you get a compile error if you try to return something from the queryFn that is either undefined or a union with undefined.

what's your use-case for having a successful query with undefined data that you cannot model with `null` instead ?

**Reply by 90dy** [reply]
It was an usage of useQuery to cache calls preventively and not run twice or more like useEffect will do, as the component was used multiple time in the same page:

```ts
const Component = () => {
  useQuery([...], () => {
    await execSomeCall1()
    await execSomeCall2()
    await execSomeCall3()
  })
  return (...)
 }
```
And I didn't understand why it didn't cache the query via the queryKey

It was a temporary code fix because the codebase is dirty.

**Reply by TkDodo** [reply]
so `return null` at the ned 🤷

**Reply by 90dy** [reply]
Yes but it's implicit

**Comment by alexian-ic** [comment]
I actually agree here that it seems odd that undefined has a reserved purpose for signaling internal cache state changes (or non-changes in this case), which means that it can no longer be used to represent the value of the cached item itself.  

My use case was to set the cache value to undefined to represent that the value is in fact no longer defined (in our case, it was deleted).

We can and likely will use null (or an empty obj) instead, but these other falsy vals have their own meaning too, so I'm wondering if there's a more graceful solution here longer term.

**Reply by TkDodo** [reply]
> My use case was to set the cache value to undefined to represent that the value is in fact no longer defined (in our case, it was deleted).

remove the cache entry with `queryClient.removeQueries`
